spec/test-utils/test-utils.ts

```diff
@@ -486,3 +486,61 @@ CRYPTO_BACKENDS["rust-sdk"] = (client: MatrixClient) => client.initRustCrypto();
 if (global.Olm) {
     CRYPTO_BACKENDS["libolm"] = (client: MatrixClient) => client.initCrypto();
 }
+
```

Comment:
I'd be more than happy to replace this with existing code or a third-party library if you know of one. Specifically, I needed to:
- imperatively resolve/reject promises
- manipulate promises that were created inside callbacks, so getting references to them would be difficult (especially given some of these callbacks would be called multiple times in the same test)

Reply:
We have a utility called `defer` which is likely to be helpful here

Reply:
perfect – got rid of a lot of this code using that, thank you [Replace new controllablePromiseFactory utility with defer + simple cache](https://github.com/matrix-org/matrix-js-sdk/pull/3530/commits/a555e70f93c6decc562d3d9f970cb4dc56869f49)